Add binary serialization support for identity-ref

dpino opened the pull request:
There are several YANG types that when compiled to binary format, they're compiled as strings. Currently, these types are: string, enumeration, empty, leafref and identityref. Instance-identifier should be also compiled to a string. The PR also implements `instance-identifier` type.

wingo commented on src/lib/yang/binary.lua:
```diff
@@ -210,11 +210,19 @@ local function data_emitter(production)
    end
+   local native_types = {
+      ["instance-identifier"] = true,
+      empty = true,
```
an empty type will have no data, no?

dpino replied:
Correct, it will have no data. Point is `empty` is implemented as a string-like datatype (it lacks ctype). That was the main reason I added to this list of "native-types". If I removed from this list, it will tried to be compiled as a ctype value and since it lacks one the code will fail.

I wrote down a test that reads a schema instance featuring an empty node and writes it down to a binary file. It's not working. I got the following error:

```
lib/yang/binary.lua:53: attempt to index a boolean value
stack traceback:
        core/main.lua:149: in function '__index'
        lib/yang/binary.lua:53: in function 'emit'
        lib/yang/binary.lua:259: in function 'compile_data_for_schema'
        test.lua:26: in function 'selftest'
        program/snsh/snsh.lua:31: in function <program/snsh/snsh.lua:31>
        core/lib.lua:434: in function 'dogetopt'
        program/snsh/snsh.lua:65: in function 'run'
        core/main.lua:67: in function <core/main.lua:43>
        [C]: in function 'xpcall'
        core/main.lua:211: in main chunk
        [C]: at 0x55b4bb2c2f30
        [C]: in function 'pcall'
        core/startup.lua:3: in main chunk
        [C]: in function 'require'
        [string "require "core.startup""]:1: in main chunk
```

This is the test:

```lua
module(..., package.seeall)

local binary = require("lib.yang.binary")
local data = require("lib.yang.data")
local schema = require("lib.yang.schema")

local compile_data_for_schema = binary.compile_data_for_schema

function selftest (args)
   local test_schema = schema.load_schema([[module test {
      namespace "test";
      prefix "test";

      container foo {
         leaf enable-qos {
            type empty;
         }
      }
   }]])
   local data = data.load_data_for_schema(test_schema, [[
      foo {
         enable-qos;
      }
   ]])
   local tmp = os.tmpname()
   compile_data_for_schema(test_schema, data, tmp)
end
```

The test gets fixed if empty.tostring returns "true", instead of an empty string. It also made me think whether an empty type should be compiled as ctype bool that always returns true. I think this makes more sense. Thoughts?

wingo commented on src/lib/yang/binary.lua:
```diff
@@ -210,11 +210,19 @@ local function data_emitter(production)
+      empty = true,
+      enumeration = true,
+      identityref = true,
+      leafref = true,
```
is the idea that leafrefs will be resolved later, somehow?  i.e. the parsed value is a name, and then later you have to look up that name

dpino replied:
In current master, `leafref` is not implemented yet. I can remove `leafref` from the list and add it later once `leafref` is mplemented. There's an attempt to implement `leafref` in https://github.com/Igalia/snabb/pull/827. It works but I think the code needs to be refactored as it's not very elegant.

dpino commented:
I was wrong, I need to add `leafref`to the list of string-like types although it's not implemented yet. Here's my use case:

I modified `snabb-softwire-v2.yang` and added several definitions from `ietf-alarms.yang` that include `leafref`nodes. When running the lwAFTR, it first compiles a schema instance to binary format. Then I got the following error:

```
$ sudo ./snabb lwaftr run --reconfigurable --name lwaftr --conf lwaftr-migrated-2.conf --on-a-stick 83:00.0
lwaftr-migrated-2.conf: loading compiled configuration from lwaftr-migrated-2.o
lwaftr-migrated-2.conf: compiled configuration is up to date.
Warning: No NUMA memory affinity.
Pass --cpu to bind to a CPU and its NUMA node.
lib/yang/binary.lua:237: unimplemented: leafref
stack traceback:
```

So I still would like o include `leafref` to the string-like types.

wingo commented on src/lib/yang/value.lua:
```diff
@@ -78,12 +78,12 @@ function types.decimal64.tostring(val)
 end
 
-types.empty = {}
+types.empty = {ctype='bool'}
```
NACK here.  No need to ever specify a value for instances of EMPTY.  Binary serialization should have a separate case for empty that doesn't serialize a value.

wingo commented on src/lib/yang/value.lua:
```diff
@@ -95,7 +95,15 @@ function types.identityref.tostring(val)
 
-types['instance-identifier'] = unimplemented('instance-identifier')
+types["instance-identifier"] = {}
+types["instance-identifier"].parse = function (str, what)
```
Can we do the instance-identifier stuff in a separate PR please?  Thanks :)

wingo reviewed (changes requested):
Would you mind restricting this PR to just add support for identityref?  I feel like we have more work to do for leafref and instance-identifier, and I'd rather get an assertion in the meantine.  If you agree, please retitle PR to "add binary serialization support for identity-ref".

dpino committed:
Add binary serialization support for identity-ref

wingo reviewed (approved):
thanks!
